contributor-book/philosophy_0_80.md

```diff
@@ -0,0 +1,182 @@
+
+Composing commands, both built-in and user-defined, is a core piece of Nushell. The design of Nushell and its standard library must support both easily composing commands as well as allowing the user to easily create compose-able commands.
+
+## Command signatures and their parts
+
+### Signature
+
+The signature of the command describes the following:
+
+- The name of the command
+- The usage information (documentation) for the command
+- The name and type of:
+    - Named parameters
+    - Positional parameters
+- The type of:
+    - Input
+    - Output
+
+### When to use input
+
+Input is intended for one or both of:
+- Pipeline composition
+- Streams of data
+
+### When to use output
+
+All return values are output.
+
+### When to use positional parameters
+
+Use a positional parameter when:
+- A parameter is required
+
+### When to use rest parameters
+
+Use a rest parameter when:
+- A command takes some number of additional optional arguments of the same type
+
+### When to use named parameters
+
+Use a named parameter (a flag parameter) when:
+- The value you need is optional
+
+### When to use a switch flag
+
+Use a switch flag when:
+- You need to allow the user to change the default action of the command
+
+### Comments
+
+Commands should be commented to describe their function and usage. This documentation should also cover parameters and their use.
+
+## Core categories
+
+The core language and standard library needs to cover the following categories to support common use cases for Nushell:
+
+* Filesytem
+* Operating system
+* Manipulating the environment
+* Parsing string data into structure data
+* Querying, filtering, and manipulating structured data
+* Network connectivity
+    * note: Network support is fundamental because with it users can easily acquire and install Nushell extensions
+* Basic formats:
+    * CSV and JSON support
+* Basic date support
+
+The following categories should be moved to plugins:
+
+* Database connectivity
+* Dataframe support
+* Hash functionality
+* Format support
+    * Any non-CSV or JSON file format
+* Experimental commands
+* Binary data functionality
+* Random number support
+* Math support
+* Advanced date support
+
+We should select commands to be in the core categories which meet the common use cases for Nushell. Commands that are in the core categories that are uncommon use cases should move to optional extensions.
+
+## Language design
+
+Rather than describe the whole of the language here, this section describes the changes expected to come as part of 0.80 and how they differ from the 0.60 series.
+
+### More bash-ism
+
+Being a shell-focused language means incorporating more of the expected shell-features into the language. These include:
+
+Redirection (with common variations):
+```
+cat foo.txt > bar.txt
+cat foo.txt >> bar.txt
+cat foo.txt 2> bar.txt
+```
+
+Bash logic operators:
+```
+cat foo.txt && cat bar.txt
+cat foo.txt || cat bar.txt
+```
```

Comment:
I definitely agree with those two categories of bashisms entering the language.

I wouldn't want to see much more though without plenty of justifying design discussion.

Reply:
Agreed - I think we want to catch the most common bash-isms without trying for some kind of full bash-compatibility.

Reply:
I'm with @dandavison partially. The redirection operators from bash are notoriously difficult to use correctly, but some thought on how to do it in a nutshell way would be useful.

It's very difficult to redirect certain things in nushell that bash can do without resorting to just jumping down into bash with a do block. Specifically things like infinite streams of stderr, or splitting stderr and stdout to different files at the same time, teeing one of them, etc

Reply:
Added some wording to limit this to just the bashisms described.